### ISSUE TITLE
Custom objective does not have the same loss curves. 

### ISSUE THREAD

**Issue by nd7141** [microsoft/LightGBM#5350]
I see that if I use custom L2 implementation instead of default one, I get much slower convergence. Why is that? I would expect two lines will be exactly the same. 

Plot: 

![image](https://user-images.githubusercontent.com/4607709/177185292-35a826c7-ae86-42f4-afc0-7ea2fe7792c2.png)

Code: 

```
X, y = sklearn.datasets.fetch_california_housing(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, train_size=0.05, random_state=0)
print(f"Train Data: {X_train.shape[0]} examples, {X_train.shape[1]} features")

params = {
    'boosting_type': 'gbdt',
    'metric': {'l2'},
}

evals_result = {}
model = lightgbm.train(params, train_data,
                       num_boost_round=200,
                       valid_sets=[train_data, test_data],
                       evals_result=evals_result,
                      )

train_rmse_default = evals_result['training']['l2']

def l2_loss(y, data):
    t = data.get_label()
    grad = y - t 
    hess = np.ones_like(y)
    return grad, hess

evals_result = {}
model = lightgbm.train(params, train_data,
                       fobj = l2_loss,
                       num_boost_round=200,
                       valid_sets=[train_data, test_data],
                       evals_result=evals_result,
                      )

train_rmse_custom = evals_result['training']['l2']

```

**Comment by jmoralez** [comment]
Hi @nd7141, I believe this is due to the different init scores, as explained in https://github.com/microsoft/LightGBM/issues/5114#issuecomment-1084994020. If you set `boost_from_average` to false with the built-in objective or set the init_score to the mean of the labels when using the custom one you should see identical results.

**Comment by jmoralez** [comment]
Yes, there's an example in https://github.com/microsoft/LightGBM/issues/5114#issuecomment-1084994020 passing it to [Dataset](https://lightgbm.readthedocs.io/en/v3.3.2/pythonapi/lightgbm.Dataset.html#lightgbm.Dataset) but you can also pass it to the [fit method](https://lightgbm.readthedocs.io/en/v3.3.2/pythonapi/lightgbm.LGBMRegressor.html#lightgbm.LGBMRegressor.fit) in the scikit-learn API.

Please let us know if you have further doubts.
